packages/expo-battery/android/src/main/java/expo/modules/battery/BatteryModule.java

```diff
@@ -0,0 +1,173 @@
+
+  private ModuleRegistry mModuleRegistry;
+  static protected Context mContext;
+  static private EventEmitter mEventEmitter;
+
+  public BatteryModule(Context context) {
+    super(context);
+    mContext = context;
+  }
+
+  public enum BatteryState {
+    CHARGING("charging"),
+    FULL("full"),
+    UNPLUGGED("unplugged"),
+    UNKNOWN("unknown");
+
+    private final String value;
+
+    BatteryState(String value) {
+      this.value = value;
+    }
+
+    public String getValue() {
+      return value;
+    }
+  }
+
+
+  @Override
+  public String getName() {
+    return NAME;
+  }
+
+  @Override
+  public void onCreate(ModuleRegistry moduleRegistry) {
+    mModuleRegistry = moduleRegistry;
+    mEventEmitter = moduleRegistry.getModule(EventEmitter.class);
+    mContext.registerReceiver(new BatteryStateReceiver(), new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
+    mContext.registerReceiver(new PowerSaverReceiver(), new IntentFilter("android.os.action.POWER_SAVE_MODE_CHANGED"));
+    IntentFilter ifilter = new IntentFilter();
+    ifilter.addAction(Intent.ACTION_BATTERY_LOW);
+    ifilter.addAction(Intent.ACTION_BATTERY_OKAY);
+    mContext.registerReceiver(new BatteryLevelReceiver(), ifilter);
+  }
+
+  static protected void onBatteryStateChange(BatteryState batteryState) {
+    Bundle result = new Bundle();
+    result.putString("batteryState", batteryState.getValue());
+    mEventEmitter.emit(BATTERY_CHARGED_EVENT_NAME, result);
+  }
+
+  static protected void onLowPowerModeChange(boolean lowPowerMode) {
+    Bundle result = new Bundle();
+    result.putString("lowPowerMode", lowPowerMode ? "on" : "off");
+    mEventEmitter.emit(POWERMODE_EVENT_NAME, result);
+  }
+
+  static protected void onBatteryLevelChange(float BatteryLevel) {
+    Bundle result = new Bundle();
+    result.putFloat("BatteryLevel", BatteryLevel);
+    mEventEmitter.emit(BATTERY_LEVEL_EVENT_NAME, result);
+  }
+
+
+  @ExpoMethod
+  public void getBatteryLevelAsync(Promise promise) {
+    try {
+      Intent batteryIntent = this.mContext.getApplicationContext().registerReceiver(null, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
+      int level = batteryIntent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);
+      int scale = batteryIntent.getIntExtra(BatteryManager.EXTRA_SCALE, -1);
+      float batteryLevel = level / (float) scale;
+      promise.resolve(batteryLevel);
+    } catch (NullPointerException e) {
+      Log.e(TAG, e.getMessage());
+      promise.reject("ERR_BATTERY", "Null Exception in trying to get battery level");
+    }
+
+  }
+
+  @ExpoMethod
+  public void getBatteryStateAsync(Promise promise) {
+    try {
+      IntentFilter ifilter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
+      Intent batteryStatus = this.mContext.getApplicationContext().registerReceiver(null, ifilter);
+      int status = batteryStatus.getIntExtra(BatteryManager.EXTRA_STATUS, -1);
+      if (status == BatteryManager.BATTERY_STATUS_FULL) {
```

Comment:
better to extract all of this logic to a separate method, maybe something like `public static BatteryState batteryStatusNativeToJS(int status)` (public so that `BatteryStateReceiver` can use it too)